freelist_test.go

```diff
@@ -44,6 +44,131 @@ func TestFreelist_release(t *testing.T) {
 	}
 }
 
+type testRange struct {
+	begin, end txid
+}
+
+type testPage struct {
+	id       pgid
+	n        int
+	allocTxn txid
+	freeTxn  txid
+}
+
+var releaseRangeTests = []struct {
+	title         string
+	pagesIn       []testPage
+	releaseRanges []testRange
+	wantFree      []pgid
+}{
+	{
+		title:         "Single pending in range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{1, 300}},
+		wantFree:      []pgid{3},
+	},
+	{
+		title:         "Single pending with minimum end range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{1, 200}},
+		wantFree:      []pgid{3},
+	},
+	{
+		title:         "Single pending outsize minimum end range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{1, 199}},
+		wantFree:      nil,
+	},
+	{
+		title:         "Single pending with minimum begin range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{100, 300}},
+		wantFree:      []pgid{3},
+	},
+	{
+		title:         "Single pending outside minimum begin range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 100, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{101, 300}},
+		wantFree:      nil,
+	},
+	{
+		title:         "Single pending in minimum range",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 199, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{199, 200}},
+		wantFree:      []pgid{3},
+	},
+	{
+		title:         "Single pending and read transaction at 199",
+		pagesIn:       []testPage{testPage{id: 3, n: 1, allocTxn: 199, freeTxn: 200}},
+		releaseRanges: []testRange{testRange{100, 198}, testRange{200, 300}},
+		wantFree:      nil,
+	},
+	{
+		title: "Adjacent pending and read transactions at 199, 200",
+		pagesIn: []testPage{
+			testPage{id: 3, n: 1, allocTxn: 199, freeTxn: 200},
+			testPage{id: 4, n: 1, allocTxn: 200, freeTxn: 201},
+		},
+		releaseRanges: []testRange{
+			testRange{100, 198},
+			testRange{200, 199},
```

Comment:
Is this range supposed to be out of order? In fact should this range even be here even correctly ordered?

Reply:
I'll add a comment. This is intended to simulate the ranges `db.freePages` might produce.